Flesh out type creation using the new TypeRegistry

ahgittin opened the pull request:
Switch many uses of Catalog to RegisteredType and improve the API around the TypePlanTransformer.  See individual commits for detail.

Next is:
* enable CAMP to use bean creation and register some beans
* swap CAMP to use a new simpler implementation for parsing, enabling reflection-based doc generation and code completion proposals
* persist RegisteredType instances instead of CatalogItems

Note that this builds on #1016 (which should be quick to review) so ignore commit  036ba79.

neykov commented on api/src/main/java/org/apache/brooklyn/api/internal/AbstractBrooklynObjectSpec.java:
```diff
@@ -160,5 +172,9 @@ public boolean equals(Object obj) {
+
+    /** strings inserted as flags, config keys inserted as config keys; 
+     * if you want to force one or the other, create a ConfigBag and convert to the appropriate map type */
+    public abstract SpecT configure(Map<?,?> val);
```
Configuration handling between specs seems identical these days, why not promote all related implementations?

But at least promote abstract `getFlags()/getConfig` as otherwise the javadoc doesn't make much sense, wdyt?

ahgittin replied:
you're right, that simplifies things -- have promoted all flag/config methods, and done related tidy

neykov commented on api/src/main/java/org/apache/brooklyn/api/mgmt/classloading/BrooklynClassLoadingContext.java:
```diff
@@ -0,0 +1,50 @@
+     * As {@link ClassLoader#getResources(String)} but returning an {@link Iterable} rather than
+     * an {@link java.util.Enumeration}.
+     */
+    public Iterable<URL> getResources(String name);
```
Not friendly to return `Iterable`, may be it's the right time to change to `Collection`?

ahgittin replied:
what's unfriendly about `Iterable`?  personally i ilke it, and it means implementations are more efficient.  (it's better than `Enumeration` which java's `ClassLoader` returns, and presumably implementations there support lazy eval.)

neykov replied:
I guess it's down to personal preference. My view is take `Iterable` and return the most specific type possible, `Iterable` is for lazy containers only. What I don't like about `Iterable` is that it's not of much use outside of guava, most libraries take `Collection`s.
The implementations are all eager in this case.

neykov commented on api/src/main/java/org/apache/brooklyn/api/typereg/BrooklynTypeRegistry.java:
```diff
@@ -45,24 +45,25 @@
-    // TODO when we support beans
-//    <T> T createBean(RegisteredType type, @Nullable RegisteredTypeConstraint optionalConstraint, Class<T> optionalResultSuperType);
+    <T> T createBean(RegisteredType type, @Nullable RegisteredTypeLoadingContext optionalConstraint, Class<T> optionalResultSuperType);
+    <T> T createBeanFromPlan(String planFormat, Object planData, @Nullable RegisteredTypeLoadingContext optionalConstraint, Class<T> optionalBeanSuperType);
```
I'd say that all the `createXXX` methods don't belong to the registry. Makes more sense if another piece of code does this, having a reference to the registry. Even `createSpec` on the CI was out of place for me, but as a single method was a 'don't bother' case.

ahgittin replied:
Interesting point.  Logically they could be moved.  But where would we move them to?  Would we have another `managementContext.getTypeInstantiator()` which uses the `TypeRegistry` ?  That doesn't seem so bad.  Let's think on this.

neykov replied:
No need for a new entry point on the context, a `new TypeInstantiator(mgmt).createBean(...)` will do.

neykov commented on api/src/main/java/org/apache/brooklyn/api/typereg/BrooklynTypeRegistry.java:
```diff
@@ -45,24 +45,25 @@
      * using a default marker if no version can be inferred */
-    RegisteredType get(String symbolicNameWithOptionalVersion, @Nullable RegisteredTypeConstraint constraint);
-    /** as {@link #get(String, RegisteredTypeConstraint)} but with no constraints */
+    RegisteredType get(String symbolicNameWithOptionalVersion, @Nullable RegisteredTypeLoadingContext constraint);
```
It's not clear from the javadoc what does the `constraint` filter on `get` methods. Do we allow conflicting types - for example a catalog item and a bean with the same `symbolicName` where `version` is not enough of a discriminator?
   * if yes, then shouldn't the return type be a collection for the case where the `constraint

neykov replied:
It's not clear from the javadoc what does the `constraint` filter on `get` methods. Do we allow conflicting types in the registry - for example a catalog item and a bean with the same `symbolicName` (in which case `version` is not enough of a discriminator)?
  * if yes, then shouldn't the return type be a collection for the case where the `constraint` wasn't strict enough?
  * if no, then why have a constraint at all here, I see it's not used yet in the implementation?

ahgittin replied:
You're right.

In general it serves a few purposes:
* it supplies class-loading context
* it helps with recursive resolution (`alreadyEncounteredType`)
* it prevents returning incompatible classes

But none of them are needed for a lookup.  Maybe we should remove; I'll look at that.

BTW correct that we don't allow different instances of the same symbolic name+version.  However i have in my mind a concept of `aliases` which types could declare, and those could conflict, so if looking up an alias the context could be useful.

neykov replied:
All three points above apply to the instantiator code path only, not the lookup.

ahgittin replied:
Yup I reached the same conclusion.  Testing your change now.

neykov commented on core/src/main/java/org/apache/brooklyn/core/plan/PlanNotRecognizedException.java:
```diff
@@ -18,8 +18,14 @@
+@Deprecated 
 public class PlanNotRecognizedException extends RuntimeException {
 
+    /** {@link UnsupportedTypePlanException} */
```
Don't see the point of the comment?

neykov commented on core/src/main/java/org/apache/brooklyn/core/plan/PlanToSpecFactory.java:
```diff
@@ -36,12 +40,30 @@
 
-    private static Collection<PlanToSpecTransformer> getAll() {
-        return ImmutableList.copyOf(ServiceLoader.load(PlanToSpecTransformer.class));
+    private static Collection<PlanToSpecTransformer> getAll(boolean includeDeprecated) {
```
The `includeDeprecated` filter doesn't have good vibes :). It's all in deprecated functionality so no big deal, but still hope it doesn't stick for too long.

ahgittin replied:
+1

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+import org.apache.brooklyn.core.mgmt.ManagementContextInjectable;
+
+/**
+ * Interface for use by schemes which with to be able to transform plans.
```
"which with" - is something missing?

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public interface BrooklynTypePlanTransformer extends ManagementContextInjectable {
+
+    /** @return a code to identify type implementations created specifying the use of this plan transformer. */
```
Sounds too twisted to me. Maybe something like "An identifier for the transformer. Can be used by `RegisteredType` instances to target a specific transformer".

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+    /** @return how appropriate is this transformer for the {@link RegisteredType#getPlan()} of the type;
+     * 0 (or less) if not, 1 for absolutely, and in some autodetect cases a value between 0 and 1 indicate a ranking.
+     * <p>
+     * The framework guarantees arguments are nonnull, and that the {@link RegisteredType#getPlan()} is also not-null.
```
Worth annotating the arguments with `@Nonnull`, more visible than the javadoc. The return values for the methods you mention as well.

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampTypePlanTransformer.java:
```diff
@@ -0,0 +1,96 @@
+
+public class CampTypePlanTransformer extends AbstractTypePlanTransformer {
+
+    private static final List<String> FORMATS = ImmutableList.of("brooklyn-camp", "camp", "brooklyn");
```
Why introduce alternative identifiers if they were never used before? It would make sense only to keep backwards compatibility, but not for new functionality.
Better stick to a single one.

ahgittin replied:
agree, for now

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+     * <p>
+     * The framework guarantees arguments are nonnull, and that the {@link RegisteredType#getPlan()} is also not-null.
+     * However the format in that plan may be null. */
+    double scoreForType(RegisteredType type, RegisteredTypeLoadingContext context);
```
Don't like this kind of non-determinism for heuristics - in particular how each implementation needs to keep an eye on the other implementations for determining the right score. Don't have a better suggestion though so will just :mute:.

ahgittin replied:
have expanded comments to try to help determine right score.  if we store formats right this won't be used very often (or possibly even ever in which case we can get rid of it)

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java:
```diff
@@ -0,0 +1,130 @@
+            }.visit(type), type, context);
+        } catch (UnsupportedTypePlanException e) {
+            // no logging
+            throw Exceptions.propagate(e);
```
My understanding is that this is just to prevent the next catch from taking over - worth a comment. May be even throwing the same exception instead of calling in to `Exceptions.propagate` to make it more visible (even if the end result is the same).

ahgittin replied:
correct.  have combined, with type check to keep same logging behaviour.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java:
```diff
@@ -0,0 +1,130 @@
+        if (type.getPlan().getPlanFormat()==null)
+            return scoreForNullFormat(type.getPlan().getPlanData(), type, context);
+        else
+            return scoreForNonmatchingNonnullFormat(type.getPlan().getPlanFormat(), type.getPlan().getPlanData(), type, context);
```
(Personal preference ahead) I am on the fail-fast side of the spectrum so my thinking is that this kind of "make every last desperate attempt to produce something usable" approach just makes implementations more complex without actually working in the wild.

ahgittin replied:
That wasn't the intention.  The idea is to simplify subclass implementations which typically do different things if the format is specified.  Have added javadoc to the class.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java:
```diff
@@ -0,0 +1,130 @@
+        return createdObject;
+    }
+
+    protected abstract AbstractBrooklynObjectSpec<?,?> createSpec(RegisteredType type, RegisteredTypeLoadingContext context) throws Exception;
```
`throws Exception` works for tests, but doesn't it mask failure points that should be handled in implementations?

ahgittin replied:
there are so many exceptions when working with classes and loading, i don't think that's likely.  happy to revert in time however; i figure this will be beta for a few releases.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BasicBrooklynTypeRegistry.java:
```diff
@@ -90,31 +96,95 @@ public RegisteredType get(String symbolicNameWithOptionalVersion, RegisteredType
+        if (result.isPresent()) return (SpecT) result.get();
         
         // fallback: look up in (legacy) catalog
+        // TODO remove once all transformers are available in the new style
```
Won't relying on the deprecated transformers make them stick around forever. Why not implement them the recommended way and be done with it?

ahgittin replied:
that's the plan.  but downstream projects may have them so keep them for at least a release or two.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BasicBrooklynTypeRegistry.java:
```diff
@@ -90,31 +96,95 @@ public RegisteredType get(String symbolicNameWithOptionalVersion, RegisteredType
+        if (item==null) {
+            // if not in catalog (because loading a new item?) then look up item based on type
+            // (only really used in tests; possibly also for any recursive legacy transformers we might have to create a CI; cross that bridge when we come to it)
+            CatalogItemType ciType = CatalogItemType.ofTargetClass( (Class)constraint.getExpectedJavaSuperType() );
```
No point in adding code to support tests, better adapt the tests?
Also don't get the recursive transformers comment, can you give an example.

ahgittin replied:
this is a port of old logic; the TODO above notes we want to remove the block

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+    Object create(RegisteredType type, RegisteredTypeLoadingContext context);
+    
+    double scoreForTypeDefinition(String formatCode, Object catalogData);
+    List<RegisteredType> createFromTypeDefinition(String formatCode, Object catalogData);
```
Will these ever be used, the registry `xxxFromPlan` methods already wrap the plans in a type?

ahgittin replied:
TBD.  My thinking is that this is where `brooklyn.catalog` syntax could be handled.  But it is very much a beta placeholder until we have implementations.

There is duplication with `BrooklynTypeRegistery.createXxxFromPlan` which will naturally be resolved as part of that work.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BrooklynTypePlanTransformer.java:
```diff
@@ -0,0 +1,68 @@
+ * <p>
+ * Implementations may wish to extend {@link AbstractTypePlanTransformer} which simplifies the process.
+ */
+public interface BrooklynTypePlanTransformer extends ManagementContextInjectable {
```
The name is confusing - it's working on `RegisteredType`s, not `BrooklynType`s?

ahgittin replied:
It's unfortunate we have `BrooklynType`.  I've proposed in comment there that *that* be renamed.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
+     * @param mgmt */
+    @Beta
+    // TODO should this be on the AbstractTypePlanTransformer ?
+    public static Class<?> loadActualJavaType(String javaTypeName, ManagementContext mgmt, RegisteredType type, RegisteredTypeLoadingContext context) throws Exception {
```
Looks like `throws Exception` is not needed.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
+    @Beta
+    // TODO should this be on the AbstractTypePlanTransformer ?
+    public static Class<?> loadActualJavaType(String javaTypeName, ManagementContext mgmt, RegisteredType type, RegisteredTypeLoadingContext context) throws Exception {
+        Class<?> result = ((BasicRegisteredType)type).getCache().get(ACTUAL_JAVA_TYPE);
```
Isn't it better to have a local `WeakHashMap` cache, than assume that the parameter will be of specific type?

ahgittin replied:
interesting idea.  have added `@Beta // TODO` on `getCache()` suggesting that, if it ends up that the cache is not wanted elsewhere

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
-        // TODO ensure this is re-populated on rebind
-        transient Class<?> javaType;
+        result = CatalogUtils.newClassLoadingContext(mgmt, type, context==null ? null : context.getLoader()).loadClass( javaTypeName );
+        Preconditions.checkNotNull(result, "Could not load class "+javaTypeName+"; returned null (should have thrown a different exception!)");
```
`loadClass` returns non null or throws - better mark it in its contract than do defensive programming at call sites.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypeLoadingContexts.java:
```diff
@@ -0,0 +1,236 @@
+            this.kind = source.getExpectedKind();
+            this.expectedSuperType = source.getExpectedJavaSuperType();
+            this.encounteredTypes = source.getAlreadyEncounteredTypes();
+            this.loader = (BrooklynClassLoadingContext) source.getLoader();
```
Cast not needed.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypeLoadingContexts.java:
```diff
@@ -0,0 +1,236 @@
+        BrooklynObjectType best = null;
+
+        for (BrooklynObjectType t: BrooklynObjectType.values()) {
+            if (t.getSpecType()==null) continue;
```
Should be `getInterfaceType()`

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypeLoadingContexts.java:
```diff
@@ -0,0 +1,236 @@
+            // else pick which is best
+            if (best==null) { best = t; continue; }
+            // if t is more specific, it is better (handles case when e.g. a Policy is a subclass of Entity)
+            if (best.getSpecType().isAssignableFrom(t.getSpecType())) { best = t; continue; }
```
Again should be `getInterfaceType()`

ahgittin replied:
good spots

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypePredicates.java:
```diff
@@ -109,36 +109,46 @@ public boolean apply(@Nullable RegisteredType item) {
 
-    public static <T> Predicate<RegisteredType> javaType(final Predicate<Class<T>> filter) {
-        return new JavaTypeMatches(filter);
+    public static <T> Predicate<RegisteredType> anySuperType(final Predicate<Class<T>> filter) {
```
Generic type is not used, replace with wildcard.

ahgittin replied:
i don't think nested generic wildcard works like that -- `Predicate<Class<?>>` does not match any `Predicate<Class<T>>`.

neykov replied:
Good point.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypePredicates.java:
```diff
@@ -109,36 +109,46 @@ public boolean apply(@Nullable RegisteredType item) {
         
         @SuppressWarnings({ "rawtypes", "unchecked" })
-        private <T> JavaTypeMatches(Predicate filter) {
+        private <T> AnySuperTypeMatches(Predicate filter) {
```
Generic type not used, remove.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypePredicates.java:
```diff
@@ -109,36 +109,46 @@ public boolean apply(@Nullable RegisteredType item) {
 
-    public static final Predicate<RegisteredType> IS_APPLICATION = javaTypeAssignableFrom(Application.class);
+    public static final Predicate<RegisteredType> IS_APPLICATION = assignableFrom(Application.class);
     // TODO do we need this?  introduced already deprecated in 0.9.0 so can be removed, or enabled
```
+1 to remove

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
-        public String getDisplayName() {
-            return displayName;
-        }
+    /** returns the implementation data for a spec if it is a string (e.g. plan yaml or java class name); else false */
```
; else null

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
+    public static String getImplementationDataStringForSpec(RegisteredType item) {
+        if (item==null || item.getPlan()==null) return null;
+        Object data = item.getPlan().getPlanData();
+        if (!(data instanceof String)) return null;
```
Looking at the usages better throw in this case.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
+    /** Returns a wrapped map, if the object is YAML which parses as a map; 
+     * otherwise returns absent capable of throwing an error with more details */
+    @SuppressWarnings("unchecked")
+    public static Maybe<Map<Object,Object>> getAsYamlMap(Object planData) {
```
Return type should be `Maybe<Map<?,?>>` (it's a read-only container), callers are not supposed to mutate it.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
+                if (isSubTypeOf((RegisteredType)st, superType)) return true;
+            }
         }
+        return false;
```
Looks the same as `RegisteredTypePredicates.assignableFrom(Class<?>)`? Re-use implementation and unify names.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypeLoadingContexts.java:
```diff
@@ -0,0 +1,236 @@
+
+public class RegisteredTypeLoadingContexts {
+
+    private static final Logger log = LoggerFactory.getLogger(RegisteredTypeLoadingContexts.BasicRegisteredTypeLoadingContext.class);
```
Weird logger, is it intentional?

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/TypePlanTransformers.java:
```diff
@@ -0,0 +1,160 @@
+    private static Collection<Class<? extends BrooklynTypePlanTransformer>> OVERRIDE;
+    @SafeVarargs
+    @VisibleForTesting
+    public synchronized static void forceAvailable(Class<? extends BrooklynTypePlanTransformer> ...classes) {
```
I like the idea of tests not using the `ServiceLoader` mechanism. Suggest passing instances instead.

ahgittin replied:
I liked the idea of instances but because instances need the management context injected that would prevent their use for rebind tests.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/TypePlanTransformers.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    public static Collection<BrooklynTypePlanTransformer> all(ManagementContext mgmt) {
+        // TODO cache these in the TypeRegistry, looking for new ones periodically or supplying a way to register them
```
Should switch to an OSGi solution instead of trying to improve this longer term.

ahgittin replied:
+1

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/TypePlanTransformers.java:
```diff
@@ -0,0 +1,160 @@
+        }
+        Map<Double, Collection<BrooklynTypePlanTransformer>> tree = new TreeMap<Double, Collection<BrooklynTypePlanTransformer>>(byScoreMulti.asMap());
+        List<Collection<BrooklynTypePlanTransformer>> highestFirst = new ArrayList<Collection<BrooklynTypePlanTransformer>>(tree.values());
+        Collections.reverse(highestFirst);
```
Could pass `Collections.reverseOrder()` to the `TreeMap` instead.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/TypePlanTransformers.java:
```diff
@@ -0,0 +1,160 @@
+        Map<Double, Collection<BrooklynTypePlanTransformer>> tree = new TreeMap<Double, Collection<BrooklynTypePlanTransformer>>(byScoreMulti.asMap());
+        List<Collection<BrooklynTypePlanTransformer>> highestFirst = new ArrayList<Collection<BrooklynTypePlanTransformer>>(tree.values());
+        Collections.reverse(highestFirst);
+        return MutableList.copyOf(Iterables.concat(highestFirst)).asUnmodifiable();
```
Could use `ImmutableList.copyOf()` instead.
Don't understand the usage of `concat` here.

ahgittin replied:
agree immutable.
concat is to flatten the (collection) entries in the treemap.
reverse as an explicit line makes the logic clearer imho.

neykov commented on core/src/test/java/org/apache/brooklyn/core/catalog/internal/StaticTypePlanTransformer.java:
```diff
@@ -0,0 +1,115 @@
+    private static final Map<String, AbstractBrooklynObjectSpec<?, ?>> REGISTERED_SPECS = new ConcurrentHashMap<>();
+
+    public static void forceInstall() {
+        TypePlanTransformers.forceAvailable(StaticTypePlanTransformer.class, JavaClassNameTypePlanTransformer.class);
```
How is this class related to `JavaClassNameTypePlanTransformer`?

ahgittin replied:
different.  this one requires the caller to provide the specs.  `JCNTPT` does a `Class.forName(typeName)` to create the spec.  added javadoc to this effect, plus a few other cleanups (clears the `REGISTERED_SPECS` map!)

neykov commented on core/src/test/java/org/apache/brooklyn/core/catalog/internal/SpecParameterInMetaTest.java:
```diff
@@ -43,18 +49,40 @@
+        Assert.assertNotNull(spec);
+    }
+
+    // it's not actually added to the catalog; probably it would be cleaner if it is;
```
Being able to create a spec from a plan is different from adding a catalog item so don't agree, it's a separate thing. The mechanism could be used for application specs as well, it's not specific to the catalog.

Could add a utility method somewhere to add a catalog item for the registered spec, but not useful for the following tests.

ahgittin replied:
I don't quite follow.  In general if a plan refers to a type, I'd expect that type in the catalog (or a java class).  While a transformer can define other rules for instantiating types, I'm not sure that's good practice.  (Except it's okay for tests.)

(have captured this discussion in the code)

neykov replied:
The idea of `TestToSpecTransformer` is to convert a plan to a pre-registered spec. It's not working on types but on plans. This means that it's able to convert a plan it understands (be it coming from a catalog item or used as an app plan) to a spec. The plan itself may be added as a catalog item, but it's a separate matter.

A plan refers to a type, but is not itself a type, it's a composition of types. Calling `registerSpec` tells the transformer how to interpret the plan, not what the spec for a specific type is.

neykov commented on core/src/test/java/org/apache/brooklyn/core/catalog/internal/SpecParameterInMetaTest.java:
```diff
@@ -43,18 +49,40 @@
+    }
+
+    // it's not actually added to the catalog; probably it would be cleaner if it is;
+    // but for now when we resolve in PlanToSpecFactory we make explicit reference to StaticTypePlanTransformer
```
Don't get this one.

neykov commented on core/src/test/java/org/apache/brooklyn/core/catalog/internal/SpecParameterInMetaTest.java:
```diff
@@ -43,18 +49,40 @@
         catalog = mgmt.getCatalog();
-        spec = TestToSpecTransformer.registerSpec(EntitySpec.create(BasicEntity.class));
+        StaticTypePlanTransformer.forceInstall();
+        PlanToSpecFactory.forceAvailable(TestToSpecTransformer.class, JavaCatalogToSpecTransformer.class);
```
Part of the tests depend on StaticTypePlanTransformer, other part on PlanToSpecFactory, but not at the same time? If so it would be nice to split them in separate files so the dependencies are clear.

ahgittin replied:
Intention is to get rid of the `TestToSpecTransformer` but we need a bit more machinery first.  Have added TODO to that effect.

neykov replied:
btw removing parameters in meta support made this class obsolete, see https://github.com/apache/incubator-brooklyn/pull/1036. Some other conflicts in the tests between the PRs unfortunately due to a renaming which was needed.

neykov commented on core/src/test/java/org/apache/brooklyn/core/catalog/internal/TestToSpecTransformer.java:
```diff
@@ -20,43 +20,36 @@
+ * <p>
+ * Should be removed when catalog is able to function using new-style lookups.
  */
 public class TestToSpecTransformer implements PlanToSpecTransformer {
```
Still can't wrap my head around the changes in this file. `StaticTypePlanTransformer` doesn't seem to be involved at all where this is used.
Looks like it's just a minimalistic CAMP parser, delegating to the registry to resolve the type. It's used in a single test - not worth it's weight (especially implementing deprecated functionality).

ahgittin replied:
It's a good exercise to make this work.  Should be cleaner when we're completely migrated.

neykov commented on core/src/test/java/org/apache/brooklyn/core/plan/XmlPlanToSpecTransformer.java:
```diff
@@ -41,6 +41,8 @@
  * how implementations are meant to be written. */
 public class XmlPlanToSpecTransformer implements PlanToSpecTransformer {
+    
+    // this is REPLACED by ExampleXmlTypePlanTransformer
```
Just delete this one?

ahgittin replied:
when we remove the PlanToSpecTransformer, yes

neykov commented on core/src/test/java/org/apache/brooklyn/core/test/BrooklynMgmtUnitTestSupport.java:
```diff
@@ -0,0 +1,58 @@
+        try {
+            if (mgmt != null) Entities.destroyAll(mgmt);
+        } catch (Throwable t) {
+            LOG.error("Caught exception in tearDown method", t);
```
I see you just copied this, but still rings an alarm in my head - aren't exceptions in tearDown part of the test failing as well? No strong opinion though.

ahgittin replied:
// we should fail here, except almost always that masks a primary failure in the test itself,
            // so it would be extremely unhelpful to do so. if we could check if test has not already failed,
            // that would be ideal, but i'm not sure if that's possible with TestNG. ?

ahgittin replied:
(agree, have added comment)

neykov commented on core/src/test/java/org/apache/brooklyn/core/typereg/JavaTypePlanTransformerTest.java:
```diff
@@ -0,0 +1,90 @@
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class JavaTypePlanTransformerTest extends BrooklynMgmtUnitTestSupport {
```
Tiny, test class name should follow class being tested.

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/api/AssemblyTemplateSpecInstantiator.java:
```diff
@@ -23,19 +23,23 @@
      * attributes and locations to the root entity.
      */
-    EntitySpec<? extends Application> createApplicationSpec(AssemblyTemplate template, CampPlatform platform, BrooklynClassLoadingContext loader);
+    EntitySpec<? extends Application> createApplicationSpec(AssemblyTemplate template, CampPlatform platform, BrooklynClassLoadingContext loader, Set<String> encounteredCatalogTypes);
```
Doesn't look right, aren't application specs always top level?

On another note, I haven't removed this method just for backwards compatibility. It belongs to an utils class with the `createServiceSpecs` doing the heavy lifting.

ahgittin replied:
agree, but the `CampResolver` methods use it, currently calling to this when they should be calilng services.  rather than repair that i hope we can deprecate all of this in favour of a new much simpler `CampTypePlanTransformer`.

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/api/AssemblyTemplateSpecInstantiator.java:
```diff
@@ -23,19 +23,23 @@
 public interface AssemblyTemplateSpecInstantiator extends AssemblyTemplateInstantiator {
 
+    @Deprecated /** @deprecaed since 0.9.0 include encountered types */
+    EntitySpec<? extends Application> createApplicationSpec(AssemblyTemplate template, CampPlatform platform, BrooklynClassLoadingContext loader);
```
No point in keeping this method. There will be compile errors anyway because of the new one and it doesn't even get called (false sense of security).

ahgittin replied:
agree

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampResolver.java:
```diff
@@ -0,0 +1,144 @@
+        AbstractBrooklynObjectSpec<?, ?> spec;
+        String planYaml = RegisteredTypes.getImplementationDataStringForSpec(item);
+        MutableSet<Object> supers = MutableSet.copyOf(item.getSuperTypes());
+        supers.addIfNotNull(expectedType);
```
Check `item` and `expectedType` compatibility?

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampResolver.java:
```diff
@@ -0,0 +1,144 @@
+
+    /** whether to allow parsing of the type syntax, where an item is a map with a "type:" field,
+     * i.e. not wrapped in any "services:" or "brooklyn.{locations,policies}" block */
+    boolean allowTypeSyntax = true;
```
Are these a part of future changes, no used at the moment?

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampResolver.java:
```diff
@@ -0,0 +1,144 @@
+            EntitySpec<? extends Application> appSpec = ((AssemblyTemplateSpecInstantiator)instantiator).createApplicationSpec(at, camp, loader, encounteredTypes);
+
+            if (!isApplication && EntityManagementUtils.canPromoteChildrenInWrappedApplication(appSpec) && appSpec.getChildren().size()==1) {
+                CampInternalUtils.resetSpecIfTemplateHasNoExplicitParameters(at, appSpec);
```
Already happened in `createWrapperApp`

ahgittin replied:
good spot

neykov commented on usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampResolver.java:
```diff
@@ -0,0 +1,144 @@
+        if (instantiator instanceof AssemblyTemplateSpecInstantiator) {
+            EntitySpec<? extends Application> appSpec = ((AssemblyTemplateSpecInstantiator)instantiator).createApplicationSpec(at, camp, loader, encounteredTypes);
+
+            if (!isApplication && EntityManagementUtils.canPromoteChildrenInWrappedApplication(appSpec) && appSpec.getChildren().size()==1) {
```
`appSpec.getChildren().size()==1` already included in the `canPromoteChildrenInWrappedApplication` check.
Also (minor) can move this block in a wrapper function instead of using a flag argument.

neykov commented on usage/camp/src/test/java/org/apache/brooklyn/camp/brooklyn/catalog/CatalogParametersTest.java:
```diff
@@ -273,7 +273,54 @@ public void testAppSpecInheritsCatalogParameters() {
+    // XXX TODO parameters on the root don't work with new type registry; 
+    // they require the CI being able to keep them,
+    // or else modifying the plan. TODO should they be permitted as metadata in this way?
+    // or treaded like a declaration of config keys on the entity?  i (alex) prefer the latter.
```
Do you plan on supporting this short term? If not remove/update mention in the catalog docs.

I have a slight preference for parameters in meta, but don't mind removing support.

ahgittin replied:
i think we've agreed we'll remove this?  we can remove the test (and this comment) when we do

neykov replied:
Yes, I commented before we did so :)

neykov commented on utils/common/src/main/java/org/apache/brooklyn/util/text/Identifiers.java:
```diff
@@ -26,6 +26,9 @@
     public static final String JAVA_GOOD_NONSTART_CHARS = JAVA_GOOD_START_CHARS+"1234567890";
+    public static final String JAVA_SEGMENT_REGEX = "["+JAVA_GOOD_START_CHARS+"]"+"["+JAVA_GOOD_NONSTART_CHARS+"]*";
+    public static final String JAVA_PACKAGE_OR_CLASS_REGEX = "("+JAVA_SEGMENT_REGEX+"\\."+")*"+JAVA_SEGMENT_REGEX;
+    public static final String JAVA_BINARY_REGEX = JAVA_PACKAGE_OR_CLASS_REGEX+"(\\$["+JAVA_GOOD_NONSTART_CHARS+"]+)*";
```
This includes a tiny subset of valid java identifiers, see http://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8. Could lead to subtle bugs.

ahgittin replied:
agree, added the word `GOOD` and javadoc

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java:
```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
-        return new RegisteredTypeKindVisitor<RegisteredTypeKind>() {
-            @Override protected RegisteredTypeKind visitSpec(RegisteredSpecType type) { return RegisteredTypeKind.SPEC; }
-        }.visit(type);
+    public static RegisteredType spec(String symbolicName, String version, TypeImplementationPlan plan, @Nullable Class<?> superType) {
```
`symbolicName`, `version` are `@Nullable` as well

ahgittin replied:
i'd like to be able to make them non-null, when the creation API is good enough

neykov commented on api/src/main/java/org/apache/brooklyn/api/typereg/BrooklynTypeRegistry.java:
```diff
@@ -45,24 +45,25 @@
     // cause compile errors, not in Eclipse, but in maven (?) 
-    <SpecT extends AbstractBrooklynObjectSpec<?,?>> SpecT createSpec(RegisteredType type, @Nullable RegisteredTypeConstraint optionalConstraint, Class<SpecT> optionalSpecSuperType);
+    <SpecT extends AbstractBrooklynObjectSpec<?,?>> SpecT createSpec(RegisteredType type, @Nullable RegisteredTypeLoadingContext optionalConstraint, Class<SpecT> optionalSpecSuperType);
+    <SpecT extends AbstractBrooklynObjectSpec<?,?>> SpecT createSpecFromPlan(String planFormat, Object planData, @Nullable RegisteredTypeLoadingContext optionalConstraint, Class<SpecT> optionalSpecSuperType);
```
`planFormat` is `@Nullable`, the optional arguments obviously as well.

neykov commented on core/src/main/java/org/apache/brooklyn/core/typereg/BasicBrooklynTypeRegistry.java:
```diff
@@ -90,31 +96,95 @@ public RegisteredType get(String symbolicNameWithOptionalVersion, RegisteredType
+        // TODO remove once all transformers are available in the new style
         CatalogItem item = (CatalogItem) mgmt.getCatalog().getCatalogItem(type.getSymbolicName(), type.getVersion());
-        return (SpecT) BasicBrooklynCatalog.internalCreateSpecWithTransformers(mgmt, item, constraint.getEncounteredTypes());
+        if (item==null) {
```
You are mixing two types of transformations in this method:
  * plan (i.e. camp) to spec transformers - `BrooklynTypePlanTransformer` (former `PlanToSpecTransformer), the part up to this line
  * type (symbolicName, java identifier) to spec transformers - `EntitySpecResolver`,  the part from this line until the end. This type of transformation is not needed at this level, it shoud be nested in the plan transformers. Looks like it's needed here just to get `SpecParameterInMetaTest` from passing.

Instead of this second part, `StaticTypePlanTransformer` should be taught how to handle the plans from the test (essentially same as pre-changes `TestToSpecTransformer`). The commit https://github.com/neykov/incubator-brooklyn/commit/8e9d69dbaeb6b15f8351f4bf5f2928c379a86b3d shows what I mean. The test `testOsgiClassScanned` is moved to `brooklyn-camp` as the camp parser is needed to create specs out of the catalog items.
With these changes the legacy transformers are used only during catalog `addItems`. Should be easy to move to the new API, I suppose part of the next PR.

ahgittin replied:
Yes, this is what I'm trying to clear up in the next PR.  Your reference is helpful!

neykov commented on core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUtils.java:
```diff
@@ -85,16 +89,23 @@ public static BrooklynClassLoadingContext getClassLoadingContext(Entity entity)
 
-        BrooklynClassLoadingContext loader = BrooklynLoaderTracker.getLoader();
-        if (loader != null) {
+        if (loader !=null) {
```
Shouldn't be allowed to stack loaders. A new loader barrier starts at a catalog item and the same loader is used for all types in the item, including specs referenced in URLs.
A loader delegates to a single item's osgi bundles, falling back to the catalog class loader. The root application loader uses only the catalog class loader.
I believe that even adding the thread local loader is wrong (in the sense that shouldn't be needed, but added just in case).

ahgittin replied:
+1 to tightening this; i've added some TODO comments and deprecation to move us in this direction

neykov commented:
Finished review, the changes look great and functionally correct. No need to reply to the copious amount of comments, just take what you think is useful.
The last two comments deserve special attention though as they touch on the high level interplay of the components.

ahgittin commented:
Awesome thorough review.  I've addressed all comments except where I'm commented.  Much better now.  Going to merge #1036 however and then clean this up once more before merging.

sjcorbett commented on api/src/main/java/org/apache/brooklyn/api/typereg/RegisteredType.java:
```diff
@@ -38,19 +46,24 @@
      */
-    Class<?> getJavaType();
+    @Beta
+    @Nullable Set<Object> getSuperTypes();
```
Any particular reason this is `@Nullable`? Feels better for it to be the opposite. Its callers already seem to assume otherwise (e.g. https://github.com/apache/incubator-brooklyn/pull/1017/files#diff-2c1cbcb42656df77efc28ab3153cb1e5R250).

ahgittin replied:
shouldn't be, will remove

ahgittin commented:
All addressed, and merged/resolved with #1036. Ready for merging if tests pass on the server.

@neykov the simpler signatures `TypeRegistry.get` are nice, and i've added validation and a simple test for the `StaticTypePlanTransformer`.  The old `Test...Transformer` is gone as I see you've moved the tests to `usage/camp` where it's not needed.
